src/core/maprenderer/qgsrendereditemresults.h

```diff
@@ -0,0 +1,89 @@
+/***************************************************************************
+  qgsrendereditemresults.h
+  -------------------
+   begin                : August 2021
+   copyright            : (C) Nyall Dawson
+   email                : nyall dot dawson at gmail dot com
+ ***************************************************************************/
+
+/***************************************************************************
+ *                                                                         *
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+
+#ifndef QGSRENDEREDITEMRESULTS_H
+#define QGSRENDEREDITEMRESULTS_H
+
+#include "qgis_core.h"
+#include "qgis_sip.h"
+#include <memory>
+#include <QList>
+#include <vector>
+
+class QgsRenderedItemDetails;
+class QgsRenderContext;
+class QgsRenderedAnnotationItemDetails;
+class QgsRectangle;
+
+///@cond PRIVATE
+class QgsRenderedItemResultsSpatialIndex;
+///@endcond
+
+/**
+ * \ingroup core
+ * \brief Stores collated details of rendered items during a map rendering operation.
+ * \since QGIS 3.22
+ */
+class CORE_EXPORT QgsRenderedItemResults
```

Comment:
Nice concept to have implemented, it'll be interesting to see what else will eventually emerge in QGIS and use this :)

Thinking out loud, it might be cool to eventually migrate the rendered labels details to this framework. Merely thinking out loud here :smile: